src/workloads/docs/StateMachineCrudActor.yml

```diff
@@ -0,0 +1,159 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/product-perf"
+
+description: |
+  TODO: This Actor does not exist yet. It will be implemented in PERF-2516.
+  The State Machine Crud Actor implements a FSM with crud operations. Choices are declared with
+  weights, which are converted into probabilities.
+
+  There is an initial state vector, with weight. Each actor instance will probabilistically and
+  independently choose it's initial state based on the weights.
+
+  Each state specifies a list of transitions. The next transition is picked probabilistically, based
+  on the transition weights. Each transition specifies a delay, thens executes a (possibly empty) set
+  of operations, before transitioning to the next state. The state itself may execute its own set of
+  operations.
+
+  By default the system will measure the latency to execute each set of operations (per transition
+  or per state), and each operation. The latencies associated with each transition and state are
+  tracked separately.
+
+  This example workload is a toy example modeling the state of a device. It has four states: On,
+  Off, Sleep, and Error, transitioning between those states. The workload does two things as it
+  executes. 1. It tracks it's current state in one document (update operations). 2. It counts the
+  number of transitions from state On to each of the other three states (also update operations).
+
+Clients:
+  Default:
+    QueryOptions:
+      maxPoolSize: 100
+
+Actors:
+# Preload some documents
+- Name: SetupDemo
+  Type: Crud
+  Threads: 1
+  Database: StateTest
+  Phases:
+  - Repeat: 1
+    Collection: CrudDemo
+    Operations:
+    - OperationName: bulkWrite
+      OperationCommand:
+        WriteOperations:
+        - WriteCommand: insertOne
+          Document: {doc: 1, state: "Off"}
+        - WriteCommand: insertOne
+          Document: {stat: "OnOff", count: 0}
+        - WriteCommand: insertOne
+          Document: {stat: "OnSleep", count: 0}
+        - WriteCommand: insertOne
+          Document: {stat: "OnError", count: 0}
```

Comment:
We want to more clearly define the responsibilities of the actor and the yaml.

In the yaml we want to define the states, the transitions, and operations to perform during each state.

But the actual FSM itself should be contained in the actor. Meaning, the initialization of the states/transitions (represented by documents), the tracking of transitions (i.e. the updateOne to the transition doc) and the actual transition itself (changing the "currentState" doc to the next state) should all be done by the actor itself. So now this section becomes:

```
Operations:
- OperationName: initFSM
   OperationCommand:
     states: ["On", "Off", "Sleep", "Error"]
```

And the state definitions look like this:
```
  - Name: StateDemo
    Type: CrudState
    Threads: 1
    Database: StateTest
    Phases:
      - *nop
      - Duration: 30 seconds
        Collection: CrudDemo
        States:
          - Name: "On"  # The device is on
            Operations:
              # A state can have (optional) a set of operations that execute upon entering the state.
              # No longer need this, the actor handles changing the "currentState" doc to "On".
              # - OperationName: updateOne
              #  OperationCommand:
              #    Filter: &filter {doc: 1}
              #    Update:
              #      $set: {state: "On"}

            # Transitions specifies next states, delays, and operations. The On state can transition to
            # each of the three other states
            Transitions:
              - To: "Off"  # Transition to this state
                Weight: 1
                # The delay is the time before firing the operations. It can be any generator. The time is
                # in seconds. In this example it will take a uniform amount of time between 0 and 15 seconds
                # to execute. The time should be a floating point number.
                Delay:
                  Delay: {^RandomDouble: {min: 0, max: 15}}
                  # You can specify the time units. Seconds is the default value and may be ommitted. Valid
                  # choices are microseconds, milliseconds, seconds, minutes, hours
                  Units: seconds
                # The actor handles creating the transition doc {start: "On", end: "Off"} and incrementing it.
                # No other ops are allowed here. 
                # Operations:
                #  - OperationName: updateOne
                #    OperationCommand:
                #      Filter: {stat: "OnOff"}
                #      Update:
                #        $inc: {count: 1}
              - To: "Sleep"
                Weight: 2
                Delay:
                  Delay: {^RandomDouble: {min: 0, max: 15}}
                # Handled by actor.
                # Operations:
                #  - OperationName: updateOne
                #    OperationCommand:
                #      Filter: {stat: "OnSleep"}
                #      Update:
                #        $inc: {count: 1}
              - To: "Error"
                Weight: 0.2
                Delay:
                  Delay: {^RandomDouble: {min: 0, max: 15}}
                # Handled by actor.
                # Operations:
                #  - OperationName: updateOne
                #    OperationCommand:
                #      Filter: {stat: "OnError"}
                #      Update:
                #        $inc: {count: 1}
```

Here's what it looks like without the comments:
```
  - Name: StateDemo
    Type: CrudState
    Threads: 1
    Database: StateTest
    Phases:
      - *nop
      - Duration: 30 seconds
        Collection: CrudDemo
        States:
          - Name: "On"  # The device is on
            Operations:
               - OperationName: RunCommand
                 # User defined command here.
            Transitions:
              - To: "Off"  # Transition to this state
                Weight: 1
                Delay:
                  Delay: {^RandomDouble: {min: 0, max: 15}}
                  Units: seconds
              - To: "Sleep"
                Weight: 2
                Delay:
                  Delay: {^RandomDouble: {min: 0, max: 15}}
              - To: "Error"
                Weight: 0.2
                Delay:
                  Delay: {^RandomDouble: {min: 0, max: 15}}
```

Reply:
The idea here is to not re-define a FSM for every workload.

Reply:
@vrachev The thought with this example was that the updateOne was part of the workload, not part of the state machine itself. It was an example of something you might want to do when you enter a state, or take a specific transition, but the actor would track its state independent of those operations. Does that make sense? 

As drawn above, what work on the cluster does the example do? I definitely would not want to hard code that work into the actor.

Reply:
Put another way -- the documents (updates  and all) are not modeling the states for the state machine. They are side-effects of the execution of the state machine. A real workload would have more interesting crud operations in their place.

Reply:
As written the operations are doing things that the state machine should be doing, so it appeared that the state machine was being implemented *in* the yaml. I just wanted to make sure we wouldn't be doing things like tracking the current state or keeping metrics on what states we've visited already. The actor should be responsible for telling us what state we're in, or how many times we've made X transition or been in Y state. Perhaps it's fine as an example, but it just confused me a bit to see that being done manually in the yaml.